### PR TITLE
Updating, fixing and improving dmd

### PR DESCRIPTION
**Pull request by ThomasMader**
This is the continuation of https://github.com/NixOS/nixpkgs/pull/11327.

It contains the following changes:
- Updated to newest version of dmd (2.075.1).
- Fetch source archives from GitHub because the archives on the dlang.org homepage have missing files. (see https://issues.dlang.org/show_bug.cgi?id=17517)
- Added a checkPhase to test the runtime and standard libraries.
- Added curl dependency because the Standard Library Phobos needs it.
- Added tzdata dependency because the datetime/timezone implementation in Phobos needs it.
- Compile with -fPIC to fix problems with binutils >= 2.28.
- Added shared library to output for Linux.
- Added manpages to output.
- Updating meta information.
- Unified the current version and bootstrap version derivation code.

- Built on platform(s)
   - [X] NixOS
   - [X] macOS

### CODE REVIEW

**Comment by joachifm** [comment]
Any idea what's up with the macos build error reported by travis?

**Comment by ThomasMader** [comment]
No wondering about it myself but Commit https://github.com/NixOS/nixpkgs/commit/740d76371e6c1c76bae4801fc5b736a796c1ebbe seems strange to me.
the referenced commit in the commit message "ba68231273bea4cba01413fd2a0e56d68db9234c" looks different and it's a really big change with 132 files.

**Comment by ThomasMader** [comment]
@joachifm For me it looks like @FRidh wanted to cherry pick one commit from staging into master but instead merged an entire branch or something.

**Comment by joachifm** [comment]
Could be intended as a partial merge to reduce parallel rebuilds.

**Comment by FRidh** [comment]
> Could be intended as a partial merge to reduce parallel rebuilds.

Exactly that.

**Comment by ThomasMader** [comment]
Nevertheless, my guess is that one of the changes on the wrapper in Commit 740d763 is the problem.

**Comment by ThomasMader** [comment]
I was able to verify that 740d763 contains the problem commit. Don't know the exact commit which broke the build of gcc-5.4.0.

@edolstra @Ericson2314 Any ideas? My guess is it's a change in the CC-wrapper stuff but don't know much about the nixpkgs structure.

I am currently building with reverted 47821f1cf0cd853d3d3dfea9259e02fea2766327 on my Mac Mini but building takes ages. :-(

**Comment by Ericson2314** [comment]
Huh, this is odd. The errors look like you are missing libc. Can you go in a nix-shell and do something like `echo '#include <stdlib.h>' | cc -E -v` and checkout out the expansion / include paths?

**Comment by ThomasMader** [comment]
`nix-shell . -A gcc5` fails with the same error as on the CI server.

nix-shell . -A clang_4
[nix-shell:~/devel/nixpkgs2]$ echo '#include <stdlib.h>' | cc -E -v
clang version 4.0.1 (tags/RELEASE_401/final)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /nix/store/bcdqax78hm8xpcbh8rilx130qrk0x91a-clang-4.0.1/bin
clang-4.0: warning: argument unused during compilation: '-idirafter /nix/store/mr8bhbnyjph6lrmak0i1bq56pyv5vg7w-Libsystem-osx-10.11.6/include' [-Wunused-command-line-argument]
clang-4.0: warning: argument unused during compilation: '-idirafter /nix/store/bcdqax78hm8xpcbh8rilx130qrk0x91a-clang-4.0.1/lib/gcc/*/*/include-fixed' [-Wunused-command-line-argument]
clang-4.0: warning: argument unused during compilation: '-F/nix/store/szvfr1ygyj7vdm9zcxizki4g3qq28xzw-CF-osx-10.10.5/Library/Frameworks' [-Wunused-command-line-argument]
clang-4.0: warning: argument unused during compilation: '-isystem /nix/store/basva09sb917d741gdn89sxljyvs0h8n-libc++-4.0.1/include' [-Wunused-command-line-argument]
clang-4.0: warning: argument unused during compilation: '-O2' [-Wunused-command-line-argument]
clang-4.0: warning: argument unused during compilation: '-D _FORTIFY_SOURCE=2' [-Wunused-command-line-argument]
clang-4.0: warning: argument unused during compilation: '-fstack-protector-strong' [-Wunused-command-line-argument]
clang-4.0: warning: argument unused during compilation: '--param ssp-buffer-size=4' [-Wunused-command-line-argument]
clang-4.0: warning: argument unused during compilation: '-fPIC' [-Wunused-command-line-argument]
clang-4.0: warning: argument unused during compilation: '-fno-strict-overflow' [-Wunused-command-line-argument]
clang-4.0: warning: argument unused during compilation: '-Wformat' [-Wunused-command-line-argument]
clang-4.0: warning: argument unused during compilation: '-Wformat-security' [-Wunused-command-line-argument]
clang-4.0: warning: argument unused during compilation: '-Werror=format-security' [-Wunused-command-line-argument]

**Comment by ThomasMader** [comment]
Here is the error message for nix-shell with gcc5:

../../gcc-5.4.0/libiberty/fibheap.c:220:30: error: use of undeclared identifier 'LONG_MIN'
  if (okey == key && okey != FIBHEAPKEY_MIN)
                             ^
../../gcc-5.4.0/libiberty/fibheap.c:38:24: note: expanded from macro 'FIBHEAPKEY_MIN'
#define FIBHEAPKEY_MIN  LONG_MIN
                        ^
../../gcc-5.4.0/libiberty/fibheap.c:261:36: error: use of undeclared identifier 'LONG_MIN'
  fibheap_replace_key (heap, node, FIBHEAPKEY_MIN);
                                   ^
../../gcc-5.4.0/libiberty/fibheap.c:38:24: note: expanded from macro 'FIBHEAPKEY_MIN'
#define FIBHEAPKEY_MIN  LONG_MIN
                        ^
../../gcc-5.4.0/libiberty/fibheap.c:265:7: warning: implicitly declaring library function 'abort' with type 'void (void) __attribute__((noreturn))' [-Wimplicit-function-declaration]
      abort ();
      ^
../../gcc-5.4.0/libiberty/fibheap.c:265:7: note: include the header <stdlib.h> or explicitly provide a declaration for 'abort'
../../gcc-5.4.0/libiberty/fibheap.c:368:3: warning: implicitly declaring library function 'memset' with type 'void *(void *, int, unsigned long)' [-Wimplicit-function-declaration]
  memset (a, 0, sizeof (fibnode_t) * D);
  ^
../../gcc-5.4.0/libiberty/fibheap.c:368:3: note: include the header <string.h> or explicitly provide a declaration for 'memset'
4 warnings and 2 errors generated.
make[3]: *** [Makefile:752: fibheap.o] Error 1
make[3]: Leaving directory '/private/tmp/nix-build-gcc-5.4.0.drv-0/build/libiberty'
make[2]: *** [Makefile:9595: all-stage1-libiberty] Error 2
make[2]: Leaving directory '/private/tmp/nix-build-gcc-5.4.0.drv-0/build'
make[1]: *** [Makefile:19039: stage1-bubble] Error 2
make[1]: Leaving directory '/private/tmp/nix-build-gcc-5.4.0.drv-0/build'
make: *** [Makefile:19344: bootstrap] Error 2
builder for ‘/nix/store/67k9dxxarnjmn80x02ix4lcwha73mbq5-gcc-5.4.0.drv’ failed with exit code 2
error: build of ‘/nix/store/67k9dxxarnjmn80x02ix4lcwha73mbq5-gcc-5.4.0.drv’ failed
/nix/var/nix/profiles/default/bin/nix-shell: failed to build all dependencies

**Comment by LnL7** [comment]
@ThomasMader FYI https://github.com/NixOS/nixpkgs/issues/27889

**Comment by ThomasMader** [comment]
I consider this done and ready to be merged.
The CI build is failing on Darwin because of https://github.com/NixOS/nixpkgs/issues/27889.
I tested on Darwin without that problem and everything is fine.

File: pkgs/development/compilers/dmd/default.nix [hunk 1]
```diff
@@ -1,44 +1,121 @@
...
+
+      # phobos uses curl, so we need to patch the path to the lib.
+      substituteInPlace phobos/posix.mak \
+          --replace "-soname=libcurl.so.4" "-soname=${curl.out}/lib/libcurl.so.4"
+  ''
+
+    + stdenv.lib.optionalString stdenv.isDarwin ''
+        # Allow to use "clang++", commented in Makefile
+        substituteInPlace dmd/posix.mak \
+            --replace g++ clang++ \
```
Comment (Ericson2314) [inline pkgs/development/compilers/dmd/default.nix]: You can replace with `$(CXX)` everywhere.

Comment (ThomasMader) [inline pkgs/development/compilers/dmd/default.nix]: Done, thanks.

**Comment by ThomasMader** [comment]
I consider this done and ready to be merged.
The CI build is failing on Darwin for the dub derivation which depends on dmd and fails because of #27889.
I tested on Darwin without that problem and everything is fine.

File: pkgs/development/compilers/dmd/2.067.1.nix [hunk 2]
```diff
@@ -1,47 +1,135 @@
...
+  buildInputs = [ curl tzdata ];
 
   # Buid and install are based on http://wiki.dlang.org/Building_DMD
   buildPhase = ''
-      cd src/dmd
+      cd dmd
       make -f posix.mak INSTALL_DIR=$out
-      export DMD=$PWD/dmd
+      ${
+          let bits = if stdenv.is64bit then "64" else "32";
```
Comment (Ericson2314) [inline pkgs/development/compilers/dmd/2.067.1.nix]: You can get that more directly from `stdenv.hostPlatform.parsed`.

File: pkgs/development/compilers/dmd/2.067.1.nix [hunk 3]
```diff
@@ -1,47 +1,135 @@
...
+          osname = if stdenv.isDarwin then "osx" else "linux"; in
```
Comment (Ericson2314) [inline pkgs/development/compilers/dmd/2.067.1.nix]: These seem unused / should be hoisted out due to repetition?

File: pkgs/development/compilers/dmd/2.067.1.nix (hunk 2 above)
Comment (ThomasMader) [inline pkgs/development/compilers/dmd/2.067.1.nix]: This is used nowhere and it isn't working for me.
So what do you mean exactly?
Using 

`let bits = stdenv.hostPlatform.parsed;`

gives me

> error: cannot coerce a set to a string

on `${bits}`.
I found the bits defined in `lib/systems/parse` but don't know how to get them out of it with your call.

Comment (Ericson2314) [inline pkgs/development/compilers/dmd/2.067.1.nix]: There's a bunch of things in that attrset. Try `stdenv.hostPlatform.parsed.cpu.bits` maybe?

Comment (Ericson2314) [inline pkgs/development/compilers/dmd/2.067.1.nix]: Oh, and in general use `nix-repl` to explore `hostPlatform` and similar values.

File: pkgs/development/compilers/dmd/2.067.1.nix (hunk 3 above)
Comment (ThomasMader) [inline pkgs/development/compilers/dmd/2.067.1.nix]: Done, thanks.

File: pkgs/development/compilers/dmd/2.067.1.nix (hunk 2 above)
Comment (ThomasMader) [inline pkgs/development/compilers/dmd/2.067.1.nix]: Done, thanks.

File: pkgs/development/compilers/dmd/default.nix [hunk 4]
```diff
@@ -47,9 +126,11 @@ stdenv.mkDerivation rec {
 
       cd ../phobos
       mkdir $out/lib
-      ${let bits = if stdenv.is64bit then "64" else "32";
-            osname = if stdenv.isDarwin then "osx" else "linux"; in
-      "cp generated/${osname}/release/${bits}/libphobos2.a $out/lib"
+      ${
+          let bits = builtins.toString stdenv.hostPlatform.parsed.cpu.bits;
+          osname = if stdenv.isDarwin then "osx" else "linux";
```
Comment (Ericson2314) [inline pkgs/development/compilers/dmd/default.nix]: Thanks for making the other two fixes! Last thing would be using `hostPlatform` and not assuming Linux:
```nix
let
  bits = builtins.toString stdenv.hostPlatform.parsed.cpu.bits;
  osname = if stdenv.hostPlatform.isDarwin then "osx" else stdenv.hostPlatform.parsed.kernel.name;
  extension = if stdenv.hostPlatform.isDarwin then "a" else "{a,so}";
in ...
```

Comment (ThomasMader) [inline pkgs/development/compilers/dmd/default.nix]: Now also fixed, thanks again.

One question. While working on those two derivations I sometimes made changes to them but `nix-build -K -A dmd` did not trigger a rebuild.
I was under the impression that changing the derivation file in any way should always lead to a new hash and therefore need to trigger a rebuild.
I then thought that maybe nix is expanding the script in any form and can therefore cancel out some changes without a rebuild.
Any way, this happend again with this last commit and I simply added two empty new lines somewhere to trigger a rebuild.

Comment (Ericson2314) [inline pkgs/development/compilers/dmd/default.nix]: Your second there is correct---if the expression evaluates to the same value, there is no rebuild. `stdenv.hostPlatform.parsed.kernel.name` evaluates to `"linux"` on Linux, so there is no change!

Thanks so much for these fixes, but don't forgot the `stdenv.hostPlatform.is*` change. `stdenv.is*` will be deprecated soon as it is ambiguous.

Comment (ThomasMader) [inline pkgs/development/compilers/dmd/default.nix]: Sorry, didn't saw your change there.
Fixed.

**Comment by FRidh** [comment]
Please squash commits

**Comment by ThomasMader** [comment]
@FRidh Squashed the small fixup commits into one.

Additionally updated to newest version 2.075.1.

**Comment by FRidh** [comment]
I still see 17 dmd commits.

**Comment by ThomasMader** [comment]
So you want me to squash all 17 commits into one?
For what reason?
Big commits are considered bad.

**Comment by ThomasMader** [comment]
@FRidh Looking through my commits, the only 2 commits which should be squashed are e35b67398308092d9c63bec40128ec5087187ed7 and e1a1ed3e1be9622f8fd8cc54c7703ba2002c2c98 .
But the others are not work-in-progress commits as mentioned in https://github.com/blog/2141-squash-your-commits .
And shouldn't you be able to squash and merge with the option explained in the blog article if you really wanted too?
But again, I really don't consider squashing all these commits into one as good pratice.

**Comment by FRidh** [comment]
> So you want me to squash all 17 commits into one?

Yes

> For what reason?
> Big commits are considered bad.

Nixpkgs is a big repository with expressions for tens of thousands of packages, not a regular repository describing a single piece of software. Navigating the history is already though and having commits for details on how an expression changes is just noise. Its important that changes are bundled, e.g. when needing to revert them.

**Comment by ThomasMader** [comment]
@FRidh Squashed everything into one commit.

Checked the build on my Linux machine again because the CI isn't working at the moment.

**Comment by ThomasMader** [comment]
> Nixpkgs is a big repository with expressions for tens of thousands of packages, not a regular repository describing a single piece of software. Navigating the history is already though and having commits for details on how an expression changes is just noise. Its important that changes are bundled, e.g. when needing to revert them.

@FRidh Wouldn't it make sense to enable the squashing option from GitHub mentioned in the blog for this repository in this case? (https://github.com/blog/2141-squash-your-commits)
This way the Merger could just decide if squashing all commits of a PR should be done or even be very strict about it and force this paradigm for all PRs independent of what the contributor has done.

**Comment by FRidh** [comment]
Yes, I can merge it here, but I cannot make sure the commit name is correct without cloning the repo.
In the end, its the responsibility of the contributor to make sure the contributions are according to the guidelines.

**Comment by Ericson2314** [comment]
@FRidh is that an official policy? I see no problem with multiple commits with a merge commit---the merge commit alone can be reverted, undoing the entire PR.

**Comment by FRidh** [comment]
@Ericson2314 we have the policy to have a commit per logical change, which typically means a commit per package or a commit per major change, not 20 commits as it was.

**Comment by Ericson2314** [comment]
@FRidh Right, that makes sense; 20 was certainly too many. But I think separating refactors from version bumps---which IIRC was some of the original motivation for multiple commits before more commits in response to reviews---is acceptable and a good idea too.

**Comment by FRidh** [comment]
If they work regardless of the version bump then absolutely.

**Comment by ThomasMader** [comment]
https://github.com/NixOS/nixpkgs/pull/28396 depends on the fixes in here.

@FRidh Is there still something missing or can it be merged?
I have created an issue for the failing CI. (see https://github.com/NixOS/nixpkgs/issues/28371)

**Comment by AlexeyRaga** [comment]
Is it fixed? Do I need to do anything special to have a fix? I am still having this error:
```
../../gcc-5.4.0/libiberty/fibheap.c:220:30: error: use of undeclared identifier 'LONG_MIN'
  if (okey == key && okey != FIBHEAPKEY_MIN)
                             ^
../../gcc-5.4.0/libiberty/fibheap.c:38:24: note: expanded from macro 'FIBHEAPKEY_MIN'
#define FIBHEAPKEY_MIN  LONG_MIN
                        ^
../../gcc-5.4.0/libiberty/fibheap.c:261:36: error: use of undeclared identifier 'LONG_MIN'
  fibheap_replace_key (heap, node, FIBHEAPKEY_MIN);
                                   ^
../../gcc-5.4.0/libiberty/fibheap.c:38:24: note: expanded from macro 'FIBHEAPKEY_MIN'
#define FIBHEAPKEY_MIN  LONG_MIN
                        ^
```

**Comment by ThomasMader** [comment]
@AlexeyRaga This was never a problem of this PR, see https://github.com/NixOS/nixpkgs/issues/27889 for details.

**Pull request merged**